xds/src/main/java/io/grpc/xds/XdsRoutingLoadBalancerProvider.java

```diff
@@ -97,77 +101,167 @@ public ConfigOrError parseLoadBalancingPolicyConfig(Map<String, ?> rawConfig) {
+      List<Route> parsedRoutes = new ArrayList<>();
+      List<Map<String, ?>> rawRoutes = JsonUtil.getListOfObjects(rawConfig, "route");
+      if (rawRoutes == null || rawRoutes.isEmpty()) {
         return ConfigOrError.fromError(Status.INTERNAL.withDescription(
             "No routes provided for xds_routing LB policy: " + rawConfig));
       }
-      List<Route> parsedRoutes = new ArrayList<>();
-      Set<MethodName> methodNames = new HashSet<>();
-      for (int i = 0; i < routes.size(); i++) {
-        Map<String, ?> route = routes.get(i);
-        String actionName = JsonUtil.getString(route, "action");
-        if (actionName == null) {
-          return ConfigOrError.fromError(Status.INTERNAL.withDescription(
-              "No action name provided for one of the routes in xds_routing LB policy: "
-                  + rawConfig));
-        }
-        if (!parsedActions.containsKey(actionName)) {
+      for (Map<String, ?> rawRoute: rawRoutes) {
+        Route route = parseRoute(rawRoute);
+        if (!parsedActions.containsKey(route.getActionName())) {
           return ConfigOrError.fromError(Status.INTERNAL.withDescription(
               "No action defined for route " + route + " in xds_routing LB policy: " + rawConfig));
         }
-        Map<String, ?> methodName = JsonUtil.getObject(route, "methodName");
-        if (methodName == null) {
-          return ConfigOrError.fromError(Status.INTERNAL.withDescription(
-              "No method_name provided for one of the routes in xds_routing LB policy: "
-                  + rawConfig));
-        }
-        String service = JsonUtil.getString(methodName, "service");
-        String method = JsonUtil.getString(methodName, "method");
-        if (service == null || method == null) {
-          return ConfigOrError.fromError(Status.INTERNAL.withDescription(
-              "No service or method provided for one of the routes in xds_routing LB policy: "
-                  + rawConfig));
+        parsedRoutes.add(route);
+      }
+      return ConfigOrError.fromConfig(new XdsRoutingConfig(parsedRoutes, parsedActions));
+    } catch (RuntimeException e) {
+      return ConfigOrError.fromError(
+          Status.fromThrowable(e).withDescription(
+              "Failed to parse xds_routing LB config: " + rawConfig));
+    }
+  }
+
+  private static PolicySelection parseAction(
+      Map<String, ?> rawAction, LoadBalancerRegistry registry) {
+    List<LbConfig> childConfigCandidates = ServiceConfigUtil.unwrapLoadBalancingConfigList(
+        JsonUtil.getListOfObjects(rawAction, "childPolicy"));
+    if (childConfigCandidates == null || childConfigCandidates.isEmpty()) {
+      throw new RuntimeException("childPolicy not specified");
+    }
+    ConfigOrError selectedConfigOrError =
+        ServiceConfigUtil.selectLbPolicyFromList(childConfigCandidates, registry);
+    if (selectedConfigOrError.getError() != null) {
+      throw selectedConfigOrError.getError().asRuntimeException();
```

Comment:
This wrapping and unwrapping create unnecessary stacktrace of exception. Some other places like
```java
} catch (RuntimeException e) {
  throw new RuntimeException("Failed to parse Route: " + e);
}
```
have the same problem

I feel the trick returning `StructOrErrorStatus<PolicySelection>` might be better.

Reply:
Slightly enhanced error propagation, threw away some stacktrace. Btw, the exception cases should ideally never happen. If it happens, the exception should have been surfaced to the Channel.

There is no perfect way to provide information about the error precisely, given that the JSON config is deeply nested and objects do not have keys that we can reference to. Unless we print out a tremendously long error message that locates the error level by level down, we are not able to point to the exact line of error in the JSON config. With that way, the error message would be something like:

```
Failed to parse xds_routing LB config: [a giant config JSON string for the whole config]
.....
Failed to parse HeaderMatcher: [a smaller but still big JSON object for the header matcher]
.....
Failed to parse Range: [a JSON object represents a Range]
....
```
This is hard to read and the effort it takes to find the malformed data is equal to proofread the whole JSON config.

The `StructOrErrorStatus` approach doesn't give better handling, it gives similar result but with verbose handling logic (at each step, always need to check if the enclosed is a struct or an error). Currently, we use a concise the error message that gives pretty much valuable hint for users to find the issue of the config.